core/trino-main/src/main/java/io/trino/operator/RowReferencePageManager.java

```diff
@@ -195,12 +195,28 @@ public boolean advance()
             return true;
         }
 
+        @Override
         public int compareTo(RowIdComparisonStrategy strategy, long rowId)
         {
             checkState(currentPosition >= 0, "Not yet advanced");
             return strategy.compare(RESERVED_ROW_ID_FOR_CURSOR, rowId);
```

Comment:
So `currentPosition` is >= 0, yet we compare `strategy.compare(RESERVED_ROW_ID_FOR_CURSOR, rowId)`. Why so?

Reply:
So this is an internal implementation detail of RowReferencePageManager that is hidden to clients:

- RowReferencePageManager allocates row IDs, and serves as the source of truth for how row IDs map to pages and positions.
- RowReferencePageManager also has at most a single active cursor to iterate through a page before generation of row IDs
- Since the active cursor only has a single page and position at any time, we need a reserved identifier so that the RowReferencePageManager knows the cursor is being referenced.

I think the key thing here is that there is only one active cursor at a time, so we can use a constant. This code used to support multiple cursors at the same time, but dain really wanted it this way since he felt it was simpler and more clear.